### PR TITLE
Add reverse proxy configuration support for remote IP address

### PR DESCRIPTION
**Pull request by lafriks**
By default trust single reverse proxy from loopback IP addresses (127.0.0.0/8, ::1/128)

In containerized environments by default trust requests from all IP addresses

EDIT: It is breaking change for users that use more then single reverse proxy to serve Gitea content

### CODE REVIEW

**Review by zeripath — APPROVED** [review]

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
how exactly is this supposed to work?
how would I set a trusted thing if Gitea listens for proxied traffic on a unix socket (I actually do that)?

**Comment by lafriks** [comment]
I think for socket remote address should be 127.0.0.1

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> I think for socket remote address should be 127.0.0.1

so they're considered as being one and the same here?

**Comment by lafriks** [comment]
I think yes, but that should probably needs to be tested

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> I think yes, but that should probably needs to be tested

would this be sort of activated by default if I build a new version?
because then tests for this could be nice prior to merge :D
if it doesn't do anything unless I have it in the config that's better but it'd still be great to have the behaviour tested and understood.

**Comment by lafriks** [comment]
> > I think yes, but that should probably needs to be tested
> 
> would this be sort of activated by default if I build a new version?
> because then tests for this could be nice prior to merge :D
> if it doesn't do anything unless I have it in the config that's better but it'd still be great to have the behaviour tested and understood.

yes, it will be activated by default otherwise it will be insecure. By default it will trust single proxy server from localhost.
Fixed library to correctly handle unix socket address that was `@` now it will be treated as `127.0.0.1:0` so trusted by default.

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -549,6 +549,12 @@ COOKIE_REMEMBER_NAME = gitea_incredible
 ; Reverse proxy authentication header name of user name
 REVERSE_PROXY_AUTHENTICATION_USER = X-WEBAUTH-USER
 REVERSE_PROXY_AUTHENTICATION_EMAIL = X-WEBAUTH-EMAIL
+; Interpret X-Forwarded-For header or the X-Real-IP header and set this as the remote IP for the request
+REVERSE_PROXY_LIMIT = 1
+; List of IP addresses seperated by comma of trusted proxy servers. Use `*` to trust all.
+REVERSE_PROXY_TRUSTED_PROXIES = 127.0.0.1
+; List of network addresses seperated by comma of trusted proxy servers. Example `10.0.0.0/24`.
+REVERSE_PROXY_TRUSTED_NETWORKS =
```
Comment (silverwind) [inline custom/conf/app.example.ini]: I think you should combine `REVERSE_PROXY_TRUSTED_PROXIES` into `REVERSE_PROXY_TRUSTED_NETWORKS`. Single IPs can be specified as `/32` (IPv4) `/128` (IPv6).

IPv4 Loopback is BTW [defined](https://en.wikipedia.org/wiki/Localhost#Name_resolution) as `127.0.0.0/8`, not a single IP which is a common misconception.

Comment (lafriks) [reply custom/conf/app.example.ini]: We could probably use single `REVERSE_PROXY_TRUSTED_PROXIES` and if string contains `/` use it as network, if not as IP address. I can change default to `127.0.0.0/8,[::1]/128`

Comment (lafriks) [reply custom/conf/app.example.ini]: @silverwind done

Comment (silverwind) [reply custom/conf/app.example.ini]: Thanks, config looks good to me now.

**Review by noerw — APPROVED** [review]

**Comment by techknowlogick** [comment]
@lafriks can I have access to your fork so I can "update branch" as needed?

**Comment by lafriks** [comment]
@techknowlogick done

**Comment by lafriks** [comment]
🚀

**Pull request merged**